Added new note type

octaedro opened the pull request:
Fixes #4656

_This is a follow up to the [PR 4621](https://github.com/woocommerce/woocommerce-admin/pull/4621)._

This PR adds a new note type `survey`. The survey note is deleted after calling its CTA.
Also, the note `WC_Admin_Notes_Insight_First_Sale.php` was modified since it belongs to this type of note. 

### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [x] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots
![Screen Capture on 2020-06-24 at 10-29-09](https://user-images.githubusercontent.com/1314156/85567094-f4b84f00-b606-11ea-8f56-cecc02623ada.gif)

### Detailed test instructions:

- Delete the notification `wc-admin-insight-first-sale`.
You can use a SQL sentence like:
````
DELETE FROM `wp_wc_admin_notes` WHERE `wp_wc_admin_notes`.`name` = 'wc-admin-insight-first-sale';
````
- Trigger the cron to add the note again. To do it you can use the plugin [WP Crontrol](https://wordpress.org/plugins/wp-crontrol/)
- Verify that the note's title is: `Did you know?`.
- Select one of the note's options.
- The message `Thanks for your feedback` should be visible.
- The note should disappear.

<!--- Note: When displaying information based on sample data, such as SwaggerHub, 
be sure to detail parts affected in Release Notes --->

### Changelog Note:
Dev: added new note type and modified the note `wc-admin-insight-first-sale` title.
<!--- Optional: Enter a changelog note following the WooCommerce core format using prefixes of Enhancement:, Tweak:, Dev:, Fix:, Performance:. If no note is entered, one will be constructed from the title and labels. --->
// cc: @pmcpinto

octaedro committed:
Added new note type

This commit adds a new note type `survey`. The note is deleted after calling its CTA.

mattsherman commented:
I'll review this

mattsherman commented on client/wc-api/constants.js:
```diff
@@ -22,5 +22,5 @@ export const QUERY_DEFAULTS = {
 	period: 'month',
 	compare: 'previous_year',
-	noteTypes: 'info,warning,marketing',
+	noteTypes: 'info,warning,marketing,survey',
```
Could we use the consts from WC_Admin_Note?

octaedro replied:
Sorry, I couldn't follow this suggestion, do you think we should load this client constant from the API?

mattsherman replied:
Ah, yeah... oops, sorry for making this suggestion. No need to load the constant from the API! Missed that this was client-side JS vs the server-side WC_Admin_Note PHP! 😊

mattsherman commented on src/API/NoteActions.php:
```diff
@@ -107,6 +107,11 @@ public function trigger_note_action( $request ) {
 		}
 
+		// Update note state depending on its type.
+		if ( 'survey' === $note->get_type() ) {
```
Could we use the const from WC_Admin_Note?

octaedro replied:
Good idea! It was done in the commit 92839f6

mattsherman commented on src/Notes/WC_Admin_Notes.php:
```diff
@@ -164,7 +164,7 @@ public static function delete_all_notes() {
 				'per_page'   => 25,
 				'page'       => 1,
-				'type'       => array( 'info', 'marketing', 'warning' ),
+				'type'       => array( 'info', 'marketing', 'warning', 'survey' ),
```
Could we use the consts from WC_Admin_Note?

octaedro replied:
Sounds good to me! I changed it in the commit 3cf623e

mattsherman commented on src/API/NoteActions.php:
```diff
@@ -107,6 +107,11 @@ public function trigger_note_action( $request ) {
 		}
 
+		// Update note state depending on its type.
+		if ( 'survey' === $note->get_type() ) {
```
This checking of type and performing some type-specific logic here is not ideal. Maybe you could hook into this using the `woocommerce_note_action` action instead, so that you could keep this deleting logic in `WC_Admin_Notes_Insight_First_Sale.php`.

octaedro replied:
I see, it's a nice catch @mattsherman! This gives me an idea😀
What do you think about creating a hook and a deletion method [in WC_Admin_Notes.php](https://github.com/woocommerce/woocommerce-admin/blob/main/src/Notes/WC_Admin_Notes.php#L25) for this type of note?
Doing that I could follow the same pattern we use to [delete the marketing notes](https://github.com/woocommerce/woocommerce-admin/blob/main/src/Notes/WC_Admin_Notes.php#L227) and we don't have to hook the deletion method in every note of this type.

mattsherman replied:
@octaedro I think I follow your idea. I think that would be good.

mattsherman reviewed (changes requested):
Tested. Works as expected, except that the entire page reloads when the note is actioned with 'Yes' or 'No' button. I would have expected the note to just hide as other notes do when you dismiss them.

Left some specific comments in-line.

octaedro committed:
Now using const from WC_Admin_Note

octaedro committed:
Now using const from WC_Admin_Note in WC_Admin_Notes too

octaedro commented:
> Tested. Works as expected, except that the entire page reloads when the note is actioned with 'Yes' or 'No' button. I would have expected the note to just hide as other notes do when you dismiss them.

Hi @mattsherman, thank you for the review 🙂
The change you mentioned makes sense. As it's something that comes from an existing notification, is it ok if we handle that change in a follow-up PR?

mattsherman commented:
> > Tested. Works as expected, except that the entire page reloads when the note is actioned with 'Yes' or 'No' button. I would have expected the note to just hide as other notes do when you dismiss them.
> 
> Hi @mattsherman, thank you for the review 🙂
> The change you mentioned makes sense. As it's something that comes from an existing notification, is it ok if we handle that change in a follow-up PR?

Sure, I think that is reasonable.

mattsherman reviewed (approved):
I reviewed the latest commits to use the constants. I didn't re-test the changes, but feel confident approving these changes since they were minor and clear.

Thanks for making those changes. Nice work!

octaedro committed:
Moved note deletion from NoteActions to WC_Admin_Notes

This commit moves the note deletion from NoteActions to WC_Admin_Notes. Also adds a hook to clean up actioned survey notes.

octaedro commented:
Hi @mattsherman,
I added the changes we mentioned in [this comment](https://github.com/woocommerce/woocommerce-admin/pull/4686#discussion_r445813423) in the commit 7caaec3.
Also, I created [another PR](https://github.com/woocommerce/woocommerce-admin/pull/4711) to handle the webpage refreshing after answering a survey.
If you have time, could you give this PR another look?

mattsherman commented on src/Notes/WC_Admin_Notes.php:
```diff
@@ -24,6 +24,7 @@ class WC_Admin_Notes {
 	 */
 	public static function init() {
 		add_action( 'admin_init', array( __CLASS__, 'schedule_unsnooze_notes' ) );
+		add_action( 'admin_init', array( __CLASS__, 'possibly_delete_survey_notes' ) );
```
Hooked into this action, the survey note will not get removed from the inbox until a full page reload. Is this what was intended?

If the note should be removed from UI after actioned, perhaps there should be a new option added when actions are registered to specify that a note should be deleted after an action is taken?

octaedro replied:
Hi @mattsherman, thank you for your review.

>Hooked into this action, the survey note will not get removed from the inbox until a full page reload. Is this what was intended?

Yes, I tried to keep the survey note, after the survey has been answered, visible with the message "Thank you for your feedback" until the next full page reload.
Maybe (just to confirm) we can ask @pmcpinto if the survey note should disappear immediately after it was activated or should stay with the message "Thank you for your feedback" until the page reload (like now).

pmcpinto commented:
> Maybe (just to confirm) we can ask @pmcpinto if the survey note should disappear immediately after it was activated or should stay with the message "Thank you for your feedback" until the page reload (like now).

It's ok to leave it until the page reload. Thanks

octaedro commented:
> It's ok to leave it until the page reload. Thanks

Thank you @pmcpinto!

@mattsherman I will add the `Needs Review` tag to this PR. Could you give it another 👀 ?

mattsherman reviewed (approved):
Thanks for the updates @octaedro -- nice work on this one!

I re-tested and verified that it works as expected.
